ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsFetch.scala

```diff
@@ -252,6 +252,54 @@ private class ContractsFetch(
     }
   }
 
+  def fetchAndRefreshCache(
+      jwt: Jwt,
+      ledgerId: LedgerApiDomain.LedgerId,
+      ledgerEnd: Terminates.AtAbsolute,
+      offsetLimitToRefresh: domain.Offset,
+  )(implicit
+      ec: ExecutionContext,
+      mat: Materializer,
+      lc: LoggingContextOf[InstanceUUID with RequestID],
+      metrics: HttpJsonApiMetrics,
+  ): ConnectionIO[Unit] = {
+    import sjd.q.queries
+    import cats.syntax.traverse._
+
+    debugLogActionWithMetrics(
+      s"cache refresh for templates older than offset: $offsetLimitToRefresh",
+      metrics.Db.warmCache,
+    ) {
+      for {
+        allOffsets <- queries.templateOffsetsOlderThan(offsetLimitToRefresh.unwrap)
+        filteredTemplateInfoAndOffset =
+          allOffsets.map { case ((packageId, moduleName, entityName), partyOffsetNonEmpty) =>
+            val partyOffset = partyOffsetNonEmpty.map { case (partyId, offset) =>
+              (domain.Party(partyId), domain.Offset(offset))
+            }
+            (ContractTypeId.Template(packageId, moduleName, entityName), partyOffset)
+          }.toList
+
+        _ <- filteredTemplateInfoAndOffset.map { case (templateId, partyOffsetNonEmpty) =>
+          val parties: domain.PartySet = partyOffsetNonEmpty.toSet.map(_._1)
+          val fetchContext = FetchContext(jwt, ledgerId, parties)
+          contractsFromOffsetIo(
+            fetchContext,
+            templateId,
+            partyOffsetNonEmpty.view.toMap,
+            true,
+            ledgerEnd,
+          )
+        }.sequence
+      } yield {
+        val templateIds = filteredTemplateInfoAndOffset.map(_._1)
+        logger.debug(
+          s"updated the cache for the follow templates: ${templateIds.mkString("[", ", ", "]")}"
+        )
+      }
```

Comment:
I think we can simplify this a bit. We can add the type information in the same pass that we actually do the refresh, with fewer conversions. I realise also that `contractsFromOffsetIo` already logs out the name of the template, so no need to logs all those names out again, but a count would probably be useful. The wrapping `debugLogActionWithMetrics` also already ensures that we log when it's done, so it's probably more worthwhile to log the count *before* attempting all the refreshes, rather than waiting for it to successfully complete.

```suggestion
      for {
        oldTemplates <- queries.templateOffsetsOlderThan(offsetLimitToRefresh.unwrap)
        _ = logger.debug(s"refreshing the cache for ${oldTemplates.size} templates")
        _ <- oldTemplates
          .map { case ((packageId, moduleName, entityName), partyOffsetsRaw) =>
            val templateId = ContractTypeId.Template(packageId, moduleName, entityName)
            val partyOffsets = partyOffsetsRaw.map { case (p, o) =>
              (domain.Party(p), domain.Offset(o))
            }.toMap
            val fetchContext = FetchContext(jwt, ledgerId, partyOffsets.keySet)
            contractsFromOffsetIo(
              fetchContext,
              templateId,
              partyOffsets,
              true,
              ledgerEnd,
            )
          }
          .toList
          .sequence
      } yield {}
```